src/testpool.jl

```diff
@@ -146,40 +137,120 @@ resize_test_engine_pool!(10, id->"RAITest-test-\$id")
+    if !isnothing(name_generator)
+        TEST_ENGINE_POOL.name_generator = name_generator
     end
 
+    @lock TEST_SERVER_LOCK begin
+        # Add engines if size > length
+        _create_and_add_engines!(size)
+        _validate_engine_pool!()
+        _trim_engine_pool!(size)
+    end
+end
+
+# Test all engines and remove if they are unavailable or not successfully provisioned
+function _validate_engine_pool!()
+    @lock TEST_SERVER_LOCK begin
+        @sync for engine in TEST_ENGINE_POOL.engines
+            try
+                response = get_engine(get_context(), engine.first; readtimeout=30)
+                if response.state == "PROVISIONED"
+                    # Success! Move on and try the next engine
+                    continue
+                end
+                # The engine exists, but is not provisioned despite our best attempts
+            catch
+                # The engine does not exist
+            end
+            # Something went wrong. Remove from the list and attempt to delete
+            delete!(TEST_ENGINE_POOL.engines, engine)
+
+            # Note that only the deletion is asynchronous, not the list modification
+            @async try
+                delete_engine(get_context(), engine.first; readtimeout=30)
+                @info("Removed failed engine $engine")
+            catch e
+                @info("Attempted to remove failed engine $engine", e)
+            end
+        end
+    end
+end
+
+function _create_and_add_engines!(size::Int64)
     @lock TEST_SERVER_LOCK begin
         engines = TEST_ENGINE_POOL.engines
+        increase = size - length(engines)
+        increase < 0 && return
+
+        new_names = String[]
         # Add engines while length < size
         while (length(engines) < size)
-            new_name = TEST_ENGINE_POOL.generator(TEST_ENGINE_POOL.next_id)
+            new_name = TEST_ENGINE_POOL.name_generator(TEST_ENGINE_POOL.next_id)
+            TEST_ENGINE_POOL.next_id += 1
+
+            # Check the engine name generator isn't repeating names.
             if haskey(engines, new_name)
                 throw(ArgumentError("Engine name already exists"))
             end
+            push!(new_names, new_name)
             engines[new_name] = 0
-            TEST_ENGINE_POOL.next_id += 1
         end
+
+        @info("Provisioning $(increase) new engines")
+        @sync for new_name in new_names
+            @async try
+                TEST_ENGINE_POOL.creater(new_name)
+            catch
+                # Ignore any errors here as we check more thoroughly below
+            end
+        end
+
+        # Test all new engines and remove if they were not successfully provisioned
+        for new_name in new_names
+            try
+                response = get_engine(get_context(), new_name; readtimeout=30)
+                if response.state == "PROVISIONED"
+                    # Success! Move on and try the next engine
+                    continue
```

Comment:
this looks similar to other code - is there an opportunity to consolidate?

Reply:
Yes, this should be handled in the validate method.